File: JavaLibrary/src/main/java/com/flat/simplex/parser/logic/block/BlockScope.java
package com.flat.simplex.parser.logic.block;

import com.flat.simplex.lexer.Key;
import com.flat.simplex.lexer.Token;
import com.flat.simplex.parser.Parser;
import com.flat.simplex.parser.logic.Block;
import com.flat.simplex.parser.logic.error.Error;

import java.util.ArrayList;

public class BlockScope extends Block {

    private Token tokenContent;
    private ArrayList<Block> blocks;

    public BlockScope(Block parent, Token start, Token end) {
        super(parent, start);

        Token token = start;
        Token lToken = start;
        int state = 0;
        while (token != end && token != null) {
            if (state == 0 && token.getKey() == Key.Brace) {
                state = 1;
                tokenContent = token;
            } else {
                error(token, Error.unexpectedToken);
            }
            lToken = token;
            token = token.getNext();
        }
        if (state < 1) {
            error(lToken, Error.unexpectedEndOfTokens);
        }
    }

    @Override
    public void read() {
        if (tokenContent != null) {
            blocks = new Parser(this).parse(tokenContent.getChild(), tokenContent.getLastChild());
            if (tokenContent.getLastChild() == null) {
                error(tokenContent, Error.missingCloser);
            }
        }
    }

    public Token getTokenContent() {
        return tokenContent;
    }
}

File: SimplexApp/simplex/lang/parser/logic/block/BlockScope.js
var Key, Parser, Error;
const Block = require("simplex/lang/parser/logic/Block.js");

class BlockScope extends Block {
    constructor(parent, start, end) {
        super(parent, start);

        this.tokenContent = null;
        this.blocks = null;

        let token = start;
        let lToken = start;
        let state = 0;
        while (token !== end && token !== null) {
            if (state === 0 && token.getKey() === Key.Brace) {
                state = 1;
                this.tokenContent = token;
            } else {
                this.error(token, Error.unexpectedToken);
            }
            lToken = token;
            token = token.getNext();
        }

        if (state < 1) {
            this.error(lToken, Error.unexpectedEndOfTokens);
        }
    }

    read() {
        if (this.tokenContent !== null) {
            this.blocks = new Parser(this).parse(this.tokenContent.getChild(), this.tokenContent.getLastChild());
            if (this.tokenContent.getLastChild() === null) {
                this.error(this.tokenContent, Error.missingCloser);
            }
        }
    }

    getTokenContent() {
        return this.tokenContent;
    }
}
module.exports = BlockScope;
Key = require("simplex/lang/lexer/Key.js");
Parser = require("simplex/lang/parser/Parser.js");
Error = require("simplex/lang/parser/logic/error/Error.js");